specs/marketplace.md

```diff
@@ -0,0 +1,374 @@
+
+1. Node detects `SlotFreed` event and decide to repair it.
+1. Node MUST download the required chunks and MUST use the [Reed-Soloman algorithm](https://hackmd.io/FB58eZQoTNm-dnhu0Y1XnA) to reconstruct the original slot's data.
+1. Node MUST generate proof over the reconstructed data.
+1. Node MUST submit transaction with call to `fillSlot()` with the same parameters and collateral allowance as described in the [Filling slot](#filling-slot).
+
+### Collecting funds
+
+A Storage Provider node SHOULD monitor Requests and slots it hosts. 
+
+When a node slot's Requests reaches states `Cancelled`, `Finished` or `Failed` it SHOULD call the contract's `freeSlot(slotId)` function.
+These states can be detected using:
+
+ - `Cancelled` state MAY be detected by setting timeout using `expiry` **and** not detecting `RequestFulfilled(requestId)` event. There is also `RequestCancelled` event emitted, but the node SHOULD NOT use it for asserting expiry as it is not guaranteed to be emitted at the time of expiry.
+ - `Finished` state MAY be detected by setting timeout specified from function `getRequestEnd(requestId)`.
+ - `Failed` state MAY be detected by listening to the `RequestFailed(requestId)` event emitted.
+
+For each of these states, different funds are collected:
+
+- For `Cancelled` the collateral is returned together with proportional payout based on time that the node actually hosted the dataset before expiry was reached.
+- For `Finished` the full reward for hosting the slot together with collateral is gathered.
+- For `Failed` no funds are collected as reward is returned to the client and collateral is burned, but this call removes the slot from the `mySlots()` tracking.
+
+## Validator role
+
+Validator role represents nodes that help to verify that the Storage Provider nodes submit proofs when they are required.  
+The decision maker around if proof was missed or not is the smart contract, validator only triggers the decision-making
+function on the smart contract. This is because in a blockchain we cannot act on things that **do not happen** and somebody needs to create a transaction 
+in order for the smart contract to act on it. The validator nodes get then rewarded for each time they correctly
+mark proof as missing.
+
+Validator nodes MUST observe the slot's space by listening on the `SlotFilled` event, which SHOULD prompt the validator
+to add the slot to the watched slots. Then after the end of every period a validator has at most `config.proofs.timeout` seconds
+(config can be retrieved with `getConfig()`) to validate all the slots and if it finds a slot that missed its proof, then 
+it SHOULD submit transaction with call to the function `markProofAsMissing(slotId, period)` that validates the correctness
+and if right, will reward the validator with a reward.
+
+Validator MAY decide to validate only part of the slot's space when it detects that it can't keep up validating slots
+before the end of validation `timeout`.
```

Comment:
```suggestion
Validator nodes MUST observe the slot's space by listening for the `SlotFilled` event, which SHOULD prompt the validator
to add the slot to the watched slots. Then after the end of every period, a validator has at most `config.proofs.timeout` seconds
(config can be retrieved with `getConfig()`) to validate all the slots. If it finds a slot that missed its proof, then 
it SHOULD submit a transaction calling the function `markProofAsMissing(slotId, period)`. This function validates the correctness of the claim,
and if right, will send a reward to the validator.

A Validator MAY decide to validate only part of the slot's space when it detects that it can't validate all watched slots
before the end of the validation `timeout`.
```